### PR TITLE
upgrade to use lodash 4.17.2

### PR DESCRIPTION
**Pull request by tchak**
---

From @mike-north:
Because lodash-4 makes use of WeakMap, where available, and because in the ember ecosystem that will come by way of babel (v6 to be specific), this is blocked by a core ember-cli issue (and really an ember-cli-babel issue)
- [ ] https://github.com/ember-cli/ember-cli/issues/5015 Babel 6.0
- [ ] https://github.com/babel/ember-cli-babel/issues/62 Babel 6.0

This isn't a trivial change to make, since we make broad use of babel in the ember ecosystem. If people want to help out, here are some upstream issues that are blocking progress on ember-cli (see [here](https://github.com/ember-cli/ember-cli/issues/5015) for the source of truth on status) 
- [x] update broccoli-babel-transpiler (https://github.com/babel/broccoli-babel-transpiler/issues/62)
- [x] update babel-plugin-feature-flags (https://github.com/ember-cli/babel-plugin-feature-flags/pull/6)
- [x] update babel-plugin-filter-imports (https://github.com/ember-cli/babel-plugin-filter-imports/pull/6)
- [x] update babel-plugin-htmlbars-inline-precompile (https://github.com/pangratz/babel-plugin-htmlbars-inline-precompile/pull/8)
- [ ] update ember-cli-htmlbars-inline-precompile ([ember-cli/ember-cli-htmlbars-inline-precompile](https://github.com/ember-cli/ember-cli-htmlbars-inline-precompile))
- [ ] update ember-cli-babel (https://github.com/babel/ember-cli-babel/issues/62)
- [ ] update emberjs-build (https://github.com/emberjs/emberjs-build/pull/165)

### CODE REVIEW

File: ember-cli-build.js [hunk 1]
```diff
@@ -5,6 +5,9 @@ var EmberAddon = require('ember-cli/lib/broccoli/ember-addon');
 module.exports = function(defaults) {
   var app = new EmberAddon(defaults, {
     // Add options here
+    babel: {
+      includePolyfill: true
```
Comment (fpauser) [inline ember-cli-build.js]: Why is this needed?

Comment (tchak) [inline ember-cli-build.js]: It should not be needed, but right now without it there is a crash in environments without native WeakMap support (like PhantomJS) :( I am trying to figure out what's happening.

Comment (tchak) [inline ember-cli-build.js]: The problem is that if WeakMap is undefined, `lodash-es` is trying to export default undefined value, and it seems that during transpilation we getting an object with default property that contains some junk instead of undefined...

Comment (fpauser) [inline ember-cli-build.js]: Ah, thanks for pointing out!

**Comment by kellyselden** [comment]
FYI I just tried out this branch, and it worked without issue. Using
`import intersectionWith from 'lodash/intersectionWith';`

**Comment by tchak** [comment]
@kellyselden is `includePolyfill` set to `true` in your projects `ember-cli-build.js`?

**Comment by kellyselden** [comment]
@tchak Nope. No babel options.

**Comment by tchak** [comment]
@kellyselden Two reasons it works for you I think. First, you using individual import. So some parts of lodash are not executed. Second, if you trying on half recent chrome or firefox, they probably support WeakMap, Map and Set. When I test with recent browsers, test only fails in PhantomJS. But I definitely do not want for the addon to surprisingly crash on older browsers...

**Comment by urbany** [comment]
Any update on this?

**Comment by Blackening999** [comment]
Updates????

File: ember-cli-build.js (hunk 1 above)
Comment (jdalton) [inline ember-cli-build.js]: Ah, I wasn't aware of this (no one brought it up). I'll patch this.

Comment (jdalton) [inline ember-cli-build.js]: I believe this is resolved in Babel 6.

Comment (tchak) [inline ember-cli-build.js]: @jdalton unfortunately ember-cli is stuck on babel 5 /cc @stefanpenner

Comment (stefanpenner) [inline ember-cli-build.js]: why am i being cc'd ?

Comment (tchak) [inline ember-cli-build.js]: @stefanpenner Sorry should have been more explicit. I rapidly mentioned to you this issue at EmberConf. The problem here is that `null`/`undefined` exports are imported as complex objects (instead of `null`/`undefined`). You said that maybe you had an idea of what the problem/fix might be.

**Comment by musaffa** [comment]
@tchak Any update?

**Comment by devinus** [comment]
Why does lodash require `WeakMap` support?

**Comment by jdalton** [comment]
Lodash uses `WeakMap` when it's available to store function wrapper metadata. The issue here is `ember-lodash` is stuck on an older buggier Babel version.

**Comment by tchak** [comment]
I must say, I ended up using `embor-browserify` for `lodash` on my current project given that even without this problem there is no `lodash/fp` es build, and I am only using `lodash/fp`...

**Comment by mike-north** [comment]
Because lodash-4 makes use of WeakMap, where available, and because in the ember ecosystem that will come by way of babel (v6 to be specific), this is blocked by a core ember-cli issue (and really an ember-cli-babel issue)

- [ ] https://github.com/ember-cli/ember-cli/issues/5015 Babel 6.0
- [ ] https://github.com/babel/ember-cli-babel/issues/62 Babel 6.0

This isn't a trivial change to make, since we make broad use of babel in the ember ecosystem. If people want to help out, here are some upstream issues that are blocking progress on ember-cli (see [here](https://github.com/ember-cli/ember-cli/issues/5015) for the source of truth on status) 

- [x] update broccoli-babel-transpiler (https://github.com/babel/broccoli-babel-transpiler/issues/62)
- [x] update babel-plugin-feature-flags (https://github.com/ember-cli/babel-plugin-feature-flags/pull/6)
- [x] update babel-plugin-filter-imports (https://github.com/ember-cli/babel-plugin-filter-imports/pull/6)
- [x] update babel-plugin-htmlbars-inline-precompile (https://github.com/pangratz/babel-plugin-htmlbars-inline-precompile/pull/8)
- [ ] update ember-cli-htmlbars-inline-precompile ([ember-cli/ember-cli-htmlbars-inline-precompile](https://github.com/ember-cli/ember-cli-htmlbars-inline-precompile))
- [ ] update ember-cli-babel (https://github.com/babel/ember-cli-babel/issues/62)
- [ ] update emberjs-build (https://github.com/emberjs/emberjs-build/pull/165)

**Comment by stefanpenner** [comment]
@jdalton both sides have bugs. v5 and v6 :( Its tricky to balance those trade-offs.

**Comment by stefanpenner** [comment]
but remember as long as your addon produces ES5 safe deanonymized AMD we don't care how you go there, if babel 6 works in your addon, it is fine to use.

**Comment by josemarluedke** [comment]
Is there anything we can do to get this rolling again? I guess I don't understand the issue here with babel 6, but is there any other alternative to make this upgrade?

**Comment by tchak** [comment]
It works! I found the way!

@mike-north should we do as planed a 3.0 and a 4.0 releases?
